### PR TITLE
Simplify AllEntries, AllField and AllFieldMapper:

### PR DESCRIPTION
**Pull request by jimczi**
- Create one AllField field per field eligible for _all.
- Add a positionIncrementGap (with a size of 100, not configurable) between
  each entry in order to distinguish fields when doing phrase query on _all.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/lucene/all/AllField.java [hunk 1]
```diff
@@ -24,60 +24,26 @@
...
-                // AllTokenStream maps boost to 4-byte payloads, so we only need to use it any field had non-default (!= 1.0f) boost and if
-                // positions are indexed:
-                return AllTokenStream.allTokenStream(name, allEntries, analyzer);
-            } else {
-                return analyzer.tokenStream(name, allEntries);
-            }
-        } catch (IOException e) {
-            throw new ElasticsearchException("Failed to create token stream");
+        TokenStream ts = analyzer.tokenStream(name(), stringValue());
+        if (boost != 1.0f && fieldType().indexOptions().compareTo(IndexOptions.DOCS_AND_FREQS_AND_POSITIONS) >= 0) {
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/common/lucene/all/AllField.java]: shouldn't we compare to DOCS_AND_FREQS_AND_POSITIONS_AND_OFFSETS? I think DOCS_AND_FREQS_AND_POSITIONS does not support payloads?

**Comment by jpountz** [comment]
This looks like a great cleanup! Does this change have any implications at runtime (I am wondering for instance if highlighting still works the same way)?

**Comment by jimferenczi** [comment]
thanks @jpountz 
Highlighting would work exactly like a multi-valued string field. Though the _all field is not stored by default (and is not stored in _source either).

**Comment by jpountz** [comment]
I played a bit with highlighting. We used to get something like this before:

```
{
  "took": 31,
  "timed_out": false,
  "_shards": {
    "total": 5,
    "successful": 5,
    "failed": 0
  },
  "hits": {
    "total": 1,
    "max_score": 0.49999997,
    "hits": [
      {
        "_index": "index",
        "_type": "t",
        "_id": "1",
        "_score": 0.49999997,
        "_source": {
          "foo": "the quick fox jumps over the lazy dog",
          "bar": "the lazy dog ate the big cat"
        },
        "highlight": {
          "_all": [
            "the quick fox jumps over the <em>lazy</em> <em>dog</em> the <em>lazy</em> <em>dog</em> ate the big cat "
          ]
        }
      }
    ]
  }
}
```

and now we have:

```
{
  "took": 59,
  "timed_out": false,
  "_shards": {
    "total": 5,
    "successful": 5,
    "failed": 0
  },
  "hits": {
    "total": 1,
    "max_score": 0.49999997,
    "hits": [
      {
        "_index": "index",
        "_type": "t",
        "_id": "1",
        "_score": 0.49999997,
        "_source": {
          "foo": "the quick fox jumps over the lazy dog",
          "bar": "the lazy dog ate the big cat"
        },
        "highlight": {
          "_all": [
            "the quick fox jumps over the <em>lazy</em> <em>dog</em>",
            "the <em>lazy</em> <em>dog</em> ate the big cat"
          ]
        }
      }
    ]
  }
}
```

which I think is much nicer.

LGTM

**Pull request merged**